ENH: loosen XLS signature

geoffrey-eisenbarth opened the pull request:
- [x] closes #41225
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them
- [x] whatsnew entry

Added ability to check for multiple XLS signatures according to testing files available at [Spreadsheet Project](https://www.openoffice.org/sc/testdocs/). Also defer raising an error when the engine is specified but the file signature does not match one of the values in `SIGNATURES`: this allows a user to attempt to specify and use an engine even if the passed file doesn't match one of the provided values in `SIGNATURES`.

This is my first PR for this project, so please let me know if more is expected (writing tests, writing a whatsnew entry, etc). I have run Flake8 on the code and successfully opened BIFF2 through BIFF8 files with this method. Thanks!

geoffrey-eisenbarth committed:
Fix #41225.

phofl reviewed (changes requested):
Please add tests

Also have to add whatsnew, but could start with tests

geoffrey-eisenbarth commented:
@phofl Thanks for the heads up. I didn't set up the pandas dev environment properly so I'm going to try to take care of that today and address the failing tests.

geoffrey-eisenbarth committed:
Adjust return type.

geoffrey-eisenbarth committed:
Update tests.

geoffrey-eisenbarth committed:
Properly stylize strings.

geoffrey-eisenbarth committed:
Correct expected exception message.

geoffrey-eisenbarth committed:
Add relevant whatsnew entry.

geoffrey-eisenbarth commented:
@phofl @rhshadrach  I think I've got it mostly squared away, however the Azure pipeline is throwing an error because it can't import `xlrd` (in order to catch `XLRDError`). Do you have a suggestion for how to resolve this? Thanks!

phofl commented:
xlrd is not installed in every ci, you should probably import this in your test runtime

phofl commented on doc/source/whatsnew/v1.2.5.rst:
```diff
@@ -34,8 +34,7 @@ Bug fixes
 ~~~~~
-
--
+- Loosen XLS signatures used in :func:`read_excel` to determine if the `xlrd` engine should be used (:issue:`41225`)
```
Please move to 1.3, I think we have a section for IO excel or something like that

phofl commented on pandas/io/excel/_base.py:
```diff
@@ -1014,16 +1015,22 @@ def close(self):
     content_or_path: FilePathOrBuffer,
     storage_options: StorageOptions = None,
-) -> str:
+) -> Optional[str]:
```
str | None instead of optional

phofl reviewed (commented):
Some cosmetic comments

geoffrey-eisenbarth commented:
> xlrd is not installed in every ci, you should probably import this in your test runtime

Just to clarify, do you mean the import should be moved to inside the testing class method? Will the test be bypassed if xlrd is not installed in the ci?

geoffrey-eisenbarth committed:
Address requested cosmetic changes.

phofl commented:
Yes into the if engine == xlrd 
If you look at the beginning of the file you will See that the engines are constructed based on the installed packages

geoffrey-eisenbarth committed:
Import exception when needed.

phofl commented on pandas/io/excel/_base.py:
```diff
@@ -1014,16 +1015,22 @@ def close(self):
     content_or_path: FilePathOrBuffer,
     storage_options: StorageOptions = None,
-) -> str:
+) -> Union[str, None]:
```
Please str | None, this is the new mechanism to type this

geoffrey-eisenbarth replied:
Oh, thanks for the heads up: I looked around in the typing documentation and didn't see this feature. Will adjust now.

phofl replied:
see https://docs.python.org/3.10/whatsnew/3.10.html#pep-604-new-type-union-operator if you are interested

geoffrey-eisenbarth committed:
Use new type hint mechanism for multiple output types.

geoffrey-eisenbarth commented:
@phofl I have updated the existing related tests, should I also write new ones that verify that `read_excel` will read older XLS files, or are we good to go?

Thanks for your help and patience!

phofl commented:
Please add new tests, if the old tests don't cover everything. Since you've added a bunch of new signatures we need test for every one of them

rhshadrach commented on pandas/io/excel/_base.py:
```diff
@@ -1037,8 +1043,8 @@ def inspect_excel_format(
-    str
-        Format of file.
+    str or None
+        Format of file (if it can be determined)
```
nit: remove parentheses - the phrase is essential to the correctness of the statement and not supplemental. Also, period at the end.

rhshadrach reviewed (changes requested):
Thanks for the PR! Looks really good, some minor requests below. Agree with @phofl on the need for tests.

geoffrey-eisenbarth commented:
> Thanks for the PR! Looks really good, some minor requests below. Agree with @phofl on the need for tests.

Thanks for the feedback! Hoping to add the tests today. I imagine it's considered bad form to write test that rely on downloading outside resources (specifically the various Excel files hosted at https://www.openoffice.org/sc/testdocs/). If that's the case, should the tests just use `io.BytesIO` objects that are created using the values in the `SIGNATURES` dictionary?

geoffrey-eisenbarth committed:
Address minor documentation recommendations.

phofl commented:
Yes, creating the tests without external files would be perfect

geoffrey-eisenbarth committed:
Add tests.

geoffrey-eisenbarth commented:
I added tests, but I'm afraid they're probably not the best. I haven't been able to find blank BIFF2, BIFF3, BIFF4, BIFF5 files in order to  observe their header and body bytes to make more accurate `io.BytesIO` streams to test against. I have examined files from the [Spreadsheet Project's Test Documents](https://www.openoffice.org/sc/testdocs/), but since those files are not empty (and I imagine some of the first few bytes describe how long the document is, etc) I'm not sure it's the best solution.

If `xlrd` throws the `struct.error` exception, then the file has passed `xlrd`'s type inspection and has been determined to be a valid BIFF file (otherwise it would throw a custom BIFF exception). I imagine the reason the `struct.error` gets thrown is because the length of the `BytesIO` does not match the length specified in the header/body bytes (although this is just speculation).

Any advice or suggestions would be greatly appreciated.

rhshadrach commented on pandas/io/excel/_base.py:
```diff
@@ -1014,16 +1014,22 @@ def close(self):
+    "biff2": b"\x09\x00\x04\x00\x07\x00\x10\x00",
+    "biff3": b"\x09\x02\x06\x00\x00\x00\x10\x00",
+    "biff4": b"\x09\x04\x06\x00\x00\x00\x10\x00",
+    "biff5": b"\xD0\xCF\x11\xE0\xA1\xB1\x1A\xE1",
+    "biff8": b"\xD0\xCF\x11\xE0\xA1\xB1\x1A\xE1",
```
Does the biff version matter? I think no - in which case can you make this a tuple with comments, e.g.

```
BIFF_SIGNATURES = (
    ...
    b"\xD0\xCF\x11\xE0\xA1\xB1\x1A\xE1",  # biff5 and biff8
)
ZIP_SIGNATURE = ...
```

rhshadrach commented on pandas/tests/io/excel/test_xlrd.py:
```diff
@@ -77,3 +77,22 @@ def test_read_excel_warning_with_xlsx_file(datapath):
+    import io
+    import struct
+
+    headers = {
```
Can you use pytest.mark.parametrize instead

rhshadrach commented on pandas/tests/io/excel/test_xlrd.py:
```diff
@@ -77,3 +77,22 @@ def test_read_excel_warning_with_xlsx_file(datapath):
+def test_read_old_xls_files():
+    # GH 41226
+    import io
+    import struct
```
Move to top of file

rhshadrach commented on pandas/tests/io/excel/test_xlrd.py:
```diff
@@ -77,3 +77,22 @@ def test_read_excel_warning_with_xlsx_file(datapath):
+    for file_format, header in headers.items():
+        f = io.BytesIO(header + body)
+        with pytest.raises(struct.error, match="unpack requires a buffer "):
+            # If struct.error is raised, file has passed xlrd's filetype checks
```
Let's just test inspect_excel_format is giving us xls for these

rhshadrach commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -831,6 +830,7 @@ I/O
 - Bug in :meth:`DataFrame.to_string` misplacing the truncation column when ``index=False`` (:issue:`40907`)
 - Bug in :func:`read_orc` always raising ``AttributeError`` (:issue:`40918`)
 - Bug in the conversion from pyarrow to pandas (e.g. for reading Parquet) with nullable dtypes and a pyarrow array whose data buffer size is not a multiple of dtype size (:issue:`40896`)
+- :func:`read_excel` now raises the specified engine's exception for incorrect file types if the excel format cannot be determined by pandas (:issue:`41225`)
```
I think the more significant (and great!) change here is that we are trying the engine even when we can't detect the file type when the user specifies it. Something like

> Bug in read_excel would raise an error when pandas could not determine the file type, even when user specified the ``engine`` argument

rhshadrach commented on pandas/tests/io/excel/test_xlrd.py:
```diff
@@ -77,3 +77,22 @@ def test_read_excel_warning_with_xlsx_file(datapath):
+        "biff4": b"\x09\x04\x06\x00\x00\x00\x10\x00",
+        "biff5": b"\xd0\xcf\x11\xe0\xa1\xb1\x1a\xe1",
+    }
+    body = b"\x00" * 16 + b"\x3e\x00\x03\x00\xfe\xff"  # Required for biff5
```
If this is required for biff5, then do we want to add that to the signatures in _base?

geoffrey-eisenbarth replied:
Upon more research, the BIFF5 (and ongoing) filetypes are actually Compound Binary Files (CFB, which have the "DOCFILE" hex header). This [US Library of Congress page](https://www.loc.gov/preservation/digital/formats/fdd/fdd000510.shtml) shows that the CFB header requires extra bits to eventually specify the endianness of the file, which is the error ``xlrd`` was giving me if I didn't include the extra ``body`` variable referenced above. However, this is no longer necessary since we're testing whether `inspect_excel_format` returns `xls` or not.

geoffrey-eisenbarth committed:
Reword based on feedback.

geoffrey-eisenbarth committed:
Refactor to test whether  returns .

geoffrey-eisenbarth committed:
Refactor to remove extraneous key information into comments.

geoffrey-eisenbarth commented on pandas/io/excel/_base.py:
```diff
@@ -1190,12 +1199,13 @@ def __init__(
                     f"only the xls format is supported. Install openpyxl instead."
                 )
-            elif ext != "xls":
+            elif ext and ext != "xls":
```
@attack68 thanks for looking over this! My understanding of the desired behavior was to go ahead and try reading it with the `xlrd` engine even if pandas can't determine the filetype (i.e., `ext=None`). I suppose I could add an `else` case to the if statement you're referencing that produces a warning?

Any input is appreciated, thanks!

attack68 replied:
maybe.. not sure what is the best solution really. just wanted to make sure the case was considered... someone else might have a good idea or just leave it with that which you stated!

rhshadrach replied:
If the user is specifying the engine explicitly, I do not think we should raise an error/warning. They have already instructed what they want to do.

geoffrey-eisenbarth commented:
@phofl @rhshadrach I believe I have addressed all the concerns (please let me know if you're still expecting anything). Should I click the "re-request review" buttons for your two requests? Thanks!

phofl commented:
Could you merge master? I'll leave the final go to @rhshadrach

geoffrey-eisenbarth committed:
Merge branch 'master' into excel-format-fixes

geoffrey-eisenbarth committed:
Address pre-commit.

geoffrey-eisenbarth committed:
Pre-commit fixes.

geoffrey-eisenbarth committed:
Merge branch 'excel-format-fixes' of https://github.com/geoffrey-eisenbarth/pandas into excel-format-fixes

rhshadrach commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -843,6 +843,8 @@ I/O
 - Bug in :func:`read_csv` and :func:`read_excel` not respecting dtype for duplicated column name when ``mangle_dupe_cols`` is set to ``True`` (:issue:`35211`)
 - Bug in :func:`read_csv` and :func:`read_table` misinterpreting arguments when ``sys.setprofile`` had been previously called (:issue:`41069`)
 - Bug in the conversion from pyarrow to pandas (e.g. for reading Parquet) with nullable dtypes and a pyarrow array whose data buffer size is not a multiple of dtype size (:issue:`40896`)
+- Bug in read_excel would raise an error when pandas could not determine the file type, even when user specified the ``engine`` argument (:issue:`41225`)
```
``:func:`read_excel` ``

rhshadrach commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -197,7 +197,7 @@ Other enhancements
 - :func:`to_numeric` now supports downcasting of nullable ``ExtensionDtype`` objects (:issue:`33013`)
 - Add support for dict-like names in :class:`MultiIndex.set_names` and :class:`MultiIndex.rename` (:issue:`20421`)
-- :func:`pandas.read_excel` can now auto detect .xlsb files (:issue:`35416`)
+- :func:`pandas.read_excel` can now auto detect .xlsb files (:issue:`35416`) and older .xls files (:issue:`41225`)
```
(:issue:`35416`, :issue:`41225`) at the end of the sentence

rhshadrach reviewed (changes requested):
lgtm, minor requests on the whatsnew.

geoffrey-eisenbarth committed:
Address requested changes in the whatsnew.

rhshadrach reviewed (approved):
lgtm

rhshadrach commented:
Thanks @geoffrey-eisenbarth!

geoffrey-eisenbarth commented:
@rhshadrach thanks for your guidance!
